### PR TITLE
cmd/bosun: expr.execute refactor

### PR DESCRIPTION
**Pull request by kylebrandt**

### CODE REVIEW

File: cmd/bosun/expr/expr.go [hunk 1]
```diff
@@ -28,31 +28,36 @@ type State struct {
...
 	// Elastic (for post ES v2)
 	elasticQueries []elastic.SearchSource
-	elasticHosts   ElasticHosts
-
-	// InfluxDB
-	InfluxConfig client.Config
+}
 
-	History AlertStatusProvider
+type Contexts struct {
```
Comment (captncraig) [inline cmd/bosun/expr/expr.go]: +1000

File: cmd/bosun/expr/expr.go [hunk 2]
```diff
@@ -85,26 +90,20 @@ func New(expr string, funcs ...map[string]parse.Func) (*Expr, error) {
 
 // Execute applies a parse expression to the specified OpenTSDB context, and
 // returns one result per group. T may be nil to ignore timings.
-func (e *Expr) Execute(c opentsdb.Context, g graphite.Context, l LogstashElasticHosts, eh ElasticHosts, influxConfig client.Config, cache *cache.Cache, T miniprofiler.Timer, now time.Time, autods int, unjoinedOk bool, search *search.Search, squelched func(tags opentsdb.TagSet) bool, history AlertStatusProvider) (r *Results, queries []opentsdb.Request, err error) {
+func (e *Expr) Execute(contexts *Contexts, cache *cache.Cache, T miniprofiler.Timer, now time.Time, autods int, unjoinedOk bool, search *search.Search, squelched func(tags opentsdb.TagSet) bool, history AlertStatusProvider) (r *Results, queries []opentsdb.Request, err error) {
```
Comment (captncraig) [inline cmd/bosun/expr/expr.go]: we still have 9 args? wow.

File: cmd/bosun/expr/expr.go (hunk 1 above)
Comment (captncraig) [inline cmd/bosun/expr/expr.go]: `Contexts` may be a bit vague. How about `Backends` ?

File: cmd/bosun/expr/expr.go (hunk 2 above)
Comment (kylebrandt) [inline cmd/bosun/expr/expr.go]: Maybe I can do better on a revision. Will look some more as long as you approve of the direction. That func signature has been making me mad for a while :P

File: cmd/bosun/expr/expr.go (hunk 1 above)
Comment (kylebrandt) [inline cmd/bosun/expr/expr.go]: K

**Comment by Dieterbe** [comment]
wasn't there plans to support multiple backends of a given type as well? this new design still looks a bit tightly coupled to particular implementations. (and singletons instances)

**Comment by kylebrandt** [comment]
nothing concrete, will be easier to do after some cleanup anyways.

**Comment by Dieterbe** [comment]
ok carry on :+1:

**Pull request merged**